CrossRefPlugin.inc.php

```diff
@@ -0,0 +1,313 @@
+     * Add properties for Crossref to the DOI entity for storage in the database.
+     *
+     * @param string $hookName `Schema::get::doi`
+     * @param array $args [
+     *      @option stdClass $schema
+     * ]
+     *
+     * @return bool
+     */
+    public function addToSchema(string $hookName, array $args): bool
+    {
+        $schema = &$args[0];
+
+        $settings = [
+            $this->_getDepositBatchIdSettingName(),
+            $this->_getFailedMsgSettingName(),
+        ];
+
+        foreach ($settings as $settingName) {
+            $schema->properties->{$settingName} = (object) [
+                'type' => 'string',
+                'apiSummary' => true,
+                'validation' => ['nullable'],
+            ];
+        }
+
+        return false;
+    }
+
+    /**
+     * Includes plugin in list of configurable registration agencies for DOI depositing functionality
+     *
+     * @param $hookName string DoiSettingsForm::setEnabledRegistrationAgencies
+     * @param $args array [
+     *      @option $enabledRegistrationAgencies array
+     * ]
+     */
+    public function addAsRegistrationAgencyOption($hookName, $args)
+    {
+        $enabledRegistrationAgencies = &$args[0];
+        $enabledRegistrationAgencies[] = [
+            'value' => $this->getName(),
+            'label' => 'Crossref'
+        ];
+    }
+
+    /**
+     * Checks if plugin meets registration agency-specific requirements for being active and handling deposits
+     *
+     * @return bool
+     */
+    public function isPluginConfigured(): bool
+    {
+        // TODO: Implement isPluginConfigured() method. For now, assume it is properly configure for testing.
+        return true;
+    }
+
+    /**
+     * Get configured registration agency display name for use in DOI management pages
+     * @return string
+     */
+    public function getRegistrationAgencyName(): string
+    {
+        return 'Crossref';
+    }
+
+    public function initiateExportAction($action, $requestBody)
+    {
+        $request = $this->getRequest(); /** @var \PKP\core\PKPRequest */
+        $context = $request->getContext();
+        $exportActionArgs = [];
+
+        if (!empty($requestBody['submissionIds'])) {
+            $exportActionArgs['submissionIds'] = $requestBody['submissionIds'];
+        }
+
+        if (!empty($requestBody['issueIds'])) {
+            $exportActionArgs['issueIds'] = $requestBody['issueIds'];
+        }
+
+        $this->_getExportPlugin()->prepareAndExportPubObjects($request, $context, $exportActionArgs);
+    }
+
+    /**
+     * Adds Crossref specific info to Repo::doi()->markRegistered()
+     *
+     * @param string $hookName Doi::markRegistered
+     * @param array $args
+     *
+     * @return bool
+     */
+    public function editMarkRegisteredParams(string $hookName, array $args): bool
+    {
+        $editParams = &$args[0];
+        $editParams[$this->_getFailedMsgSettingName()] = null;
+
+        return false;
+    }
```

Comment:
My thinking had been to use this to remove any previous Crossrgf-specific failure messages when an item is marked registered. Do you think it would be best to keep this here, or do something like remove all settings associated with a DOI when its marked registered?

Reply:
What's the reasoning for removing the information? My first instinct is to preserve it but I don't really know the context.

Reply:
My thought had been that if something previously had an error, but it's then marked as registered if the error is corrected outside of OJS, it may be confusing to continue to include the error if the user has marked that as resolved by marking it as registered. I believe the error messages are always visible if there is one, regardless of DOI state, but I'd have to double check. 

Would it be better in general to preserve the Crossref error message, or any other previously stored messages, and display it only if the DOI has an error status? That way if a user needs to refer back to it later, it's still there, but it won't appear on the UI if the DOI has been marked as registered.

Reply:
Hmm, yeah this is maybe worth some consideration. My sense is that the error message is a record of failure, rather than its current status. That's why I hesitate to clear it out. Can the UI maybe make this clearer? I'm thinking where we show the error message we could indicate something like:

> The following status was received from Crossref when this a request to deposit this DOI was made on YYYY-MM-DD.
>
> {$errorMessage}

Presumably the DOI's status would still be sent to registered and so it would appear in the list as a registered DOI.

Reply:
That makes sense. I like that approach as it preserve the history of the message while still making it clear the it's a message from a specific point in time and not necessarily reflective of the DOIs current status. I'll give that a try and see how it looks.